Improve simulation UX

survived opened the pull request:
Writing tests for MPC protocols involves copy-pasting the same boilerplate over and over again. This PR improves ergonomics by moving the most common logic into the lib. 

You can see how it improves the tests in https://github.com/LFDT-Lockness/cggmp21/pull/120, specifically in [81191c0](https://github.com/LFDT-Lockness/cggmp21/pull/120/commits/81191c0659bb9bf1f4ef901e3d6cc428fc45eb22):
- [x] A lot less code required:
  * No need to manually poll futures via `futures::future::try_join_all`
  * No need to write a check each time that all outputs are equally the same (e.g. in signing, we need to check that all sigs are the same, and we used to write a loop each time, now we just use `.expect_eq()`)
- [x] No need of async runtime. All tests are sync now. 

Other changes:
* Remove `std` feature. All errors implement `core::error::Error` trait.

survived committed:
Rework simulation

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Bump version & update changelog

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
clippy fix

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Shorten function names

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Use state machine based simulation by default

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Elaborate on changes, and add migration guidelines in changelog

Signed-off-by: Denis Varlakov <denis@dfns.co>

maurges commented on round-based/CHANGELOG.md:
```diff
@@ -1,3 +1,27 @@
+* Take advantage of `SimResult::{expect_ok, expect_eq}` to reduce amount of the code
+  in your tests
+
+Rather than simulation, there's no breaking changes in this release.
```
s/rather/other/

maurges commented on round-based/Cargo.toml:
```diff
@@ -32,11 +32,17 @@ trybuild = "1"
 default = ["std"]
 state-machine = []
-dev = ["std", "tokio/sync", "tokio-stream"]
+# dev = ["std", "tokio/sync", "tokio-stream", "futures-util/alloc"]
```
Forgot the comment?

maurges commented on round-based/src/simulation/mod.rs:
```diff
@@ -1,49 +1,496 @@
+                Outputs :\n",
+            );
+
+            for (i, res) in self.0.iter().enumerate() {
+                msg += &alloc::format!("- Party {i}: {res:?}");
+            }
```
Might be helpful to cluster results by equality, so it would be like `- Parties 1, 2: {res1:?} \n - Parties 3, 4: {res2:?}`

survived replied:
Good point! Done

maurges commented on round-based/src/simulation/mod.rs:
```diff
@@ -1,49 +1,496 @@
+        }
+    }
+
+    /// Constructs a simulation with `n` parties from async function that defines the protocol
+    ///
+    /// Each party has index `0 <= i < n` and instantiated via provided `init` function
```
Probably should have a disclaimer that this runs with our own runtime instead of tokio

maurges commented on round-based/src/simulation/mod.rs:
```diff
@@ -1,49 +1,496 @@
+    /// Constructs a simulation with `n` parties from async function that defines the protocol
+    ///
+    /// Each party has index `0 <= i < n` and instantiated via provided `init` function
+    pub fn from_async_fn<F>(
```
Probably should have a disclaimer that this runs with our own runtime instead of tokio

survived replied:
Added clarifications

survived committed:
Update docs

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Update changelog

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Remove redundant comment

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Improve UX of `expect_eq`

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Clarify docs on `Simulation::from_async_fn`

Signed-off-by: Denis Varlakov <denis@dfns.co>

maurges reviewed (approved):
One part of me wants to say that we should keep the old api, on the other hand I'm not going to use it anywhere.

One nitpick is that I'd prefer using "simulation" instead of "sim"

survived committed:
Add Simulation::add_async_party

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived commented:
> One part of me wants to say that we should keep the old api

How do we improve UX and evolve the lib if we commit to keep old api? :smiley:

survived commented:
> One nitpick is that I'd prefer using "simulation" instead of "sim"

I also thought about it, though writing "simulation" wasn't an issue when I was updating all the dependent lib... I think I'll go ahead and use shorter name

survived committed:
Rename simulation mod to `sim`, improve SimResult

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Fix docs

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Update docs

Signed-off-by: Denis Varlakov <denis@dfns.co>

maurges commented on round-based/src/state_machine/mod.rs:
```diff
@@ -4,6 +4,33 @@
+//! let state_machine = round_based::state_machine::wrap_protocol(
+//!     |party| protocol_of_random_generation(party, 0, 3)
+//! );
+//! // `state_machine` implements `round_based::state_machine::StateMachine` trait.
+//! // Its methods can be used to advance protocol until completion.
```
Maybe add an example of polling to completion? Maybe it should be in StateMachine itself, and here we can write "see example there"

survived replied:
That's a bit difficult to do, as what "advancing until completion" looks like depends on how it's used, e.g. how incoming/outgoing messages are handled, etc. I could write something like:

```rust
let mut state = state_machine;

loop {
    match state.proceed() {
        ProceedResult::SendMsg(msg) => outgoings.send(msg).await?,
        ProceedResult::NeedsOneMoreMessage => {
            let msg = incomings.next().await?;
            state.received_msg(msg)?;
        }
        ProceedResult::Yielded => {},
        ProceedResult::Output(out) => break Ok(out),
        ProceedResult::Error(err) => break Err(err),
    }
}
```

but that's a bit weird, as using async stream/sink defeats the purpose of the state machines

survived replied:
On other hand, it might be a good thing to have this example anyway, to know what to do with the state machine in general...

maurges replied:
Exactly my thoughts, state machine polling is a difficult api and right now completely lacks examples

survived replied:
Yeah true, I'll add this example

survived replied:
Added an example. Lmk if you think it can be improved @maurges

survived committed:
Add example in state machine docs

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Use `core::error::Error` trait

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived commented:
@maurges that's a bit out of scope of the PR, but I also removed uses of `std::error::Error` trait, and replaced with `core::error::Error` in 5d53357

survived committed:
Remove `std` feature

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Fix broken things

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Require that `sim` is no_std compatible

Signed-off-by: Denis Varlakov <denis@dfns.co>

maurges commented on round-based/src/state_machine/mod.rs:
```diff
@@ -25,11 +27,41 @@
-//! // `state_machine` implements `round_based::state_machine::StateMachine` trait.
-//! // Its methods can be used to advance protocol until completion.
+//!
+//! // Note: this is just an example. If you have stream/sink, you don't probably need to
+//! // use the sync API
+//! use futures::{Sink, Stream, SinkExt, StreamExt};
+//! async fn connect() -> Result<(
+//!     impl Stream<Item = anyhow::Result<round_based::Incoming<Msg>>>,
+//!     impl Sink<round_based::Outgoing<Msg>, Error = Error>
+//! )> {
+//!     // ...
+//!     # Ok((futures_util::stream::pending(), futures_util::sink::drain().sink_map_err(|err| match err {})))
+//! }
+//! let (mut incomings, mut outgoings) = connect().await?;
```
Maybe replace the async containers with simple sync functions `send(msg)` and `recv() -> msg`?

survived replied:
Yeah good idea

survived replied:
Done

survived committed:
Use sync functions instead of async in state_machine examples

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived committed:
Make clippy happy

Signed-off-by: Denis Varlakov <denis@dfns.co>

survived commented:
Too soon! @maurges Clippy suddenly started complaining on unmodified part of the library

survived commented:
That's a large PR...
